### PR TITLE
BUGFIX: Solver Refactor - Allow no objective

### PR DESCRIPTION
**Pull request by mrmundt**
## Summary/Motivation:
There was a bug reported in https://github.com/Pyomo/pyomo/issues/2574#issuecomment-1983781290 that points out that you can't use `ipopt_v2` to solve something without an objective, which you technically can do in `v1`. This fixes that bug.

## Changes proposed in this PR:
- Allow no objective 
- Populate the number of objectives / constraints / params attributes on `problem`

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Review by blnicho — APPROVED** [review]

File: pyomo/contrib/solver/base.py [hunk 1]
```diff
@@ -418,9 +418,14 @@ def _map_results(self, model, results):
...
         legacy_soln.status = legacy_solution_status_map[results.solution_status]
         legacy_results.solver.termination_message = str(results.termination_condition)
-        obj = get_objective(model)
-        if len(list(obj)) > 0:
+        legacy_results.problem.number_of_constraints = model.nconstraints()
+        legacy_results.problem.number_of_variables = model.nvariables()
+        if model.nobjectives() == 0:
+            legacy_results.problem.number_of_objectives = 0
+        else:
+            obj = get_objective(model)
```
Comment (michaelbynum) [inline pyomo/contrib/solver/base.py:426]: This function does one of three things:
1. returns `None` if there is not an active objective
2. raises an error if there are multiple active objectives
3. returns the objective if there is exactly one objective

As a result, you will not be able to do `len(obj)` below.

**Review by michaelbynum — CHANGES_REQUESTED** [review]
Thanks for fixing this @mrmundt. I left one comment below.

File: pyomo/contrib/solver/base.py (hunk 1 above)
Comment (mrmundt) [reply pyomo/contrib/solver/base.py:426]: So I should just do `model.nobjectives()` is how I interpret that. Accurate?

Comment (michaelbynum) [reply pyomo/contrib/solver/base.py:426]: Yes. Then you just need to grab the objective to get the sense.

Comment (mrmundt) [reply pyomo/contrib/solver/base.py:427]: Updated!

**Review by michaelbynum — APPROVED** [review]

File: pyomo/contrib/solver/base.py (hunk 1 above)
Comment (michaelbynum) [reply pyomo/contrib/solver/base.py:427]: yes

Comment (jsiirola) [reply pyomo/contrib/solver/base.py:427]: So this patch will raise an `AttributeError` for multiobjective NL files (which is in theory supported by the ASL / Pyomo API), in line 428 `None.sense`

Comment (mrmundt) [reply pyomo/contrib/solver/base.py:427]: What is the right way to get `sense` for multiple objectives?

Comment (jsiirola) [reply pyomo/contrib/solver/base.py:427]: A single `sense` is meaningless for multiple objectives (it should be a vector).  But the old interface never really handled MO problems well.  That said, my concern is that this is probably not the right place for a MO problem to end up with an error.

**Review by jsiirola — APPROVED** [review]

**Pull request merged**